keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,662 @@
+
+If a lien is set to an owner object, only a deletion request to the owner object is blocked.
+Therefore, a deletion request to each dependent object won't be blocked.
+In other words, users can still delete each dependent object directly, but they can't delete it through an owner object.
+
+If a lien is set to one of dependent objects, only a deletion request to the dependent object is blocked.
+Therefore, a deletion request to its owner object or the other dependent objects won't be blocked.
+In other words, users can delete its owner object or other dependent objects.
+The point that needs to be considered is whether the owner object is actually deleted, or whether its deletion completes immediately.
+It depends on the cascading policy and the dependent object's `blockOwnerDeletion`.
+The behaviors are as follows:
+
+- Foreground cascading deletion:
+  - `blockOwnerDeletion=true`: Deletion of the owner object isn't completed until the dependent object is deleted
+  - otherwise: The owner object is deleted immediately
+- Background cascading deletion: The owner object is deleted immediately
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Therefore, a deletion request to each object in the namespace won't be blocked.
+In other words, users can still delete each object in the namespace directly, but they can't delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked.
+In other words, users can delete its namespace or other objects in the namespace.
+The point that needs to be considered is whether the namespace is actually deleted, or whether its deletion completes immediately.
+On namespace deletion, all the objects in the namespace are deleted, and the deletion of the namespace isn't completed until all the objects are deleted.
+As a result, deletion of the namespace isn't completed until the object is deleted.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
+
+#### Per Controller `Liens` vs Per Reason `Liens`
```

Comment:
I'm unclear - if I interpret @detiber correctly (and if not, sorry) he's describing a dep-graph, where deletion of a lien-holder is sufficient to clear a lien.

That's SIGNIFICANTLY more involved (I think) than a simple lien.

Am I wrong?

Reply:
The ideal scenario I had in mind would indeed require a dependency graph, but I still think it could be approximated using a common controller or shared library as @mkimuram describes.

We've seen various iterations of this within the Cluster API project for a few different types of scenarios.

The main scenario we've hit is ensuring proper cleanup of resources and centralized reporting of status.  We don't want folks to have to know the proper order of deletion of specific resources to ensure a cluster is cleanly deleted, so we encourage folks to delete the Cluster resource rather than the other related resources, but it would be nice to have a way to avoid users potentially inadvertently getting things wedged by deleting things in the wrong order or accidentally deleting something they didn't mean to. Today we accomplish this mostly through documentation, and implementing deletion of "child" resources through the higher level controllers and try to ensure that we don't introduce false assumptions and bugs if deletion order is done "improperly" or if foreground deletion is used

The other use case we've hit is around introducing safety around managed child resources. Take the example of a KubeadmControlPlane, this resource manages multiple Machine and related resources that make up the control plane of the kubernetes clusters being managed. If a user unintentionally deletes managed "child" resources that "belong" to a KubeadmControlPlane, then they could potentially not only destroy their cluster, but destroy the underlying data it contained. However, if the user is already deleting the KubeadmControlPlane resource, then there is no need to provide that level of protection. KubeadmControlPlane is the obvious case we have here for Cluster API, but similar behavior could be beneficial to other "parent"/"child" CRD/controller relationships as well.